streams/src/main/java/org/apache/kafka/streams/processor/internals/RecordCollectorImpl.java

```diff
@@ -264,15 +266,23 @@ private void checkForException() {
     @Override
     public void flush() {
         log.debug("Flushing producer");
-        producer.flush();
+        try {
+            producer.flush();
+        } catch (final ProducerFencedException | UnknownProducerIdException e) {
```

Comment:
Thanks. One thing I'm not sure about, but felt at the time of putting this together was that, if we get an UnknownProducerIdException on one call, we'd probably just keep getting it on all subsequent calls. It seemed to me that the exception was originating from a condition where the broker "forgot" about our id, and if it has "forgotten" about us, I'm not sure why it would "remember" us later on.

If that's not true, then this approach certainly may result in more close+reopen operations than is necessary.